### PR TITLE
Add padding option to Popup class 

### LINKED ISSUE #5978: Introduce padding for Popup
## User Story
 As a user I can pass `popupPadding` config to `Popup` so that I can limit it's positioning with custom bounds

## Rationale
There was a [similar feature request](https://github.com/mapbox/mapbox-gl-js/pull/12708) made in mapbox in the past, and I require such a feature in my application. Implementing something like that using current maplibre popups requires overriding default behaviour, and using internal functions. I managed to make it somehow work, but the code feels hacky and relies on copying the existing maplibre popup position calculations, so adding it internally would be much nicer. 

Here is an example of how this functionality would work:
Current behaviour:

https://private-user-images.githubusercontent.com/34219675/237748834-020d8de5-4794-4d66-a707-a75e024cf822.mov?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkwNzY2MDAsIm5iZiI6MTc0OTA3NjMwMCwicGF0aCI6Ii8zNDIxOTY3NS8yMzc3NDg4MzQtMDIwZDhkZTUtNDc5NC00ZDY2LWE3MDctYTc1ZTAyNGNmODIyLm1vdj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA0VDIyMzE0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViMzA0MGYyYjMzMDEzYzk1ZmM5YzZkNjc2OTBkMzU2OTkxZTVmNGZkN2EyMzA4YjczYzQxZjcxNTMzODliOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ar7NNsUQE51LFkHCV6ARxmZObDW7Uy2_Tk0b4-R7T3c

Desired behaviour:

https://private-user-images.githubusercontent.com/34219675/237748895-3b29bf9e-8347-408b-9744-6ae1a78e8137.mov?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkwNzY2MDAsIm5iZiI6MTc0OTA3NjMwMCwicGF0aCI6Ii8zNDIxOTY3NS8yMzc3NDg4OTUtM2IyOWJmOWUtODM0Ny00MDhiLTk3NDQtNmFlMWE3OGU4MTM3Lm1vdj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA0VDIyMzE0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZGZmNzdkODhiNjQzOTIzOTg3Yjk4OWFjZmRhYzY5MzJlNzczMTM5MmZkMzc0Yzg4ZjM3YmE1ZmMyN2FjNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.blg5w4qHN74b3NsfCql0a0Hv9AnLTTNSIOHtkvFkwI8

## Impact
If implemented as an additional parameter e.g. `options.popupPadding` strictly related to popups it does not have any impact. It would be an extension that users can use as needed. Otherwise, if we rely on existing map padding, it might change the behaviour for existing applications (which I would rather avoid) 

## Implementation
I'm open and happy to add this feature myself, but would like to get your opinion on that first, and if it's a wanted feature, discuss the implementation approach.

### PR DESCRIPTION
**Pull request by lucaswoj**
This is a continuation of #6052 by @yuiseki 

cc @mizmay 

Dec-18-2025 15-21-46

> ## Launch Checklist
> * [x]  Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
> * [x]  Briefly describe the changes in this PR.
> * [x]  Link to related issues.
> * [x]  Include before/after visuals or gifs if this PR includes visual changes.
> * [x]  Write tests for all new functionality.
> * [x]  Document any changes to public APIs.
> * [x]  Post benchmark scores.
> * [x]  Add an entry to `CHANGELOG.md` under the `## main` section.
> 
> ## Summary
> This PR implements the `popupPadding` feature requested in issue #5978, allowing developers to define buffer zones around map container edges where popups should avoid being positioned. This is particularly useful for preventing popups from being hidden behind UI overlays like headers, sidebars, or toolbars.
> 
> **Closes #5978**
> 
> ## Changes Made
> **New API Features:**
> 
> * Added `popupPadding` option to `PopupOptions` interface supporting multiple input formats:
>   
>   * Number: Equal padding on all sides
>   * PointLike/Array: Horizontal and vertical padding
>   * 4-element array: CSS-style [top, right, bottom, left] padding
>   * Object: Explicit {top, right, bottom, left} properties
> * Added `setPopupPadding(padding?: PopupPadding)` method for dynamic updates
> * Added `PopupPadding` type definition with comprehensive format support
> 
> **Implementation Details:**
> 
> * Enhanced automatic anchor selection logic in `_update()` method to respect padding constraints
> * Added `normalizePopupPadding()` function with robust input validation and NaN/Infinity handling
> * Maintained complete backward compatibility - existing code unaffected
> * Manual anchor selection bypasses padding calculations (preserves existing behavior)
> * Padding constraints work seamlessly with existing offset system
> 
> ## Testing
> **Comprehensive test coverage with 90 test cases including:**
> 
> **Core Functionality:**
> 
> * Multiple padding input formats (number, object, array)
> * Dynamic padding updates via `setPopupPadding()`
> * Interaction with existing offset system
> 
> **Critical Compatibility Tests:**
> 
> * Backward compatibility verification across 9 different map positions
> * Manual anchor behavior preservation (all 9 anchor positions tested)
> * TrackPointer functionality remains unchanged
> * Performance impact assessment showing no degradation
> 
> **Edge Case Handling:**
> 
> * NaN, Infinity, negative values, and extreme numbers
> * Mixed problematic values in object/array formats
> * Zero padding vs undefined padding equivalence
> 
> **Integration Tests:**
> 
> * Offset + padding interaction validation
> * CSS positioning and styling integrity
> * API consistency with MapLibre patterns
> 
> ## Performance
> Performance testing with 100 iterations shows **no performance degradation** - actually measured -9.8% overhead (slight improvement), likely due to measurement variance. The padding calculation adds minimal computational cost.
> 
> ## API Design Rationale
> The API follows established MapLibre patterns:
> 
> * Naming convention matches existing options like `closeOnClick`, `subpixelPositioning`
> * Setter method follows `set + PascalCase` pattern like `setOffset()`, `setMaxWidth()`
> * Type flexibility mirrors other geometric properties in the codebase
> * Fluent interface support for method chaining
> 
> ## Example Usage
> ```ts
> // Basic usage - equal padding on all sides
> const popup = new Popup({popupPadding: 20});
> 
> // Custom padding for UI overlays
> const popup = new Popup({
>   popupPadding: {top: 60, left: 200} // Header + sidebar
> });
> 
> // Dynamic updates
> popup.setPopupPadding([20, 10]); // [horizontal, vertical]
> 
> // CSS-style padding
> popup.setPopupPadding([10, 20, 30, 40]); // [top, right, bottom, left]
> ```
> 
> Backward Compatibility
> 
> Zero breaking changes confirmed:
> 
> * All existing popup code continues to work unchanged
> * Default popupPadding: undefined maintains original behavior
> * Manual anchor settings completely unaffected
> * Offset calculations preserve existing logic
> 
> This feature addresses the exact use case described in the original issue where popups would appear behind headers or sidebars, providing a clean, MapLibre-native solution for constraining popup positioning.
> 
> Thank you for the thoughtful feedback on this feature request and for maintaining such high standards for MapLibre GL JS. The guidance about keeping this separate from global map padding and ensuring individual popup control was invaluable. This implementation prioritizes robustness and backward compatibility that MapLibre deserves. 🙏

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] ~Post benchmark scores.~
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/ui/popup.ts [hunk 1]
```diff
@@ -567,6 +576,21 @@ export class Popup extends Evented {
...
+     * @param padding - The padding to apply as a {@link PaddingOptions} object.
+     * @example
+     * ```ts
+     * popup.setPopupPadding({ top: 10, right: 20, bottom: 30, left: 40 });
+     * ```
+     */
+    setPopupPadding(padding?: PaddingOptions): this {
+        this.options.popupPadding = padding;
+        this._update();
+        return this;
```
Comment (HarelM) [inline src/ui/popup.ts]: While other method might return "this" I found it problematic more than once, I think it will be better to return void.

File: src/ui/popup.ts [hunk 2]
```diff
@@ -95,6 +97,13 @@ export type PopupOptions = {
      * @defaultValue undefined
      */
     locationOccludedOpacity?: number | string;
+    /**
+     * A pixel padding applied to the popup's positioning constraints.
+     * The popup will be positioned to avoid being placed within this padding area
+     * from the edges of the map container.
+     * @defaultValue undefined
+     */
+    popupPadding?: PaddingOptions;
```
Comment (HarelM) [inline src/ui/popup.ts]: I think it should be called just "padding", like other option (offset for example).

File: src/ui/popup.ts [hunk 3]
```diff
@@ -567,6 +576,21 @@ export class Popup extends Evented {
...
+    /**
+     * Sets the popup's padding constraints for positioning.
+     *
+     * @param padding - The padding to apply as a {@link PaddingOptions} object.
+     * @example
+     * ```ts
+     * popup.setPopupPadding({ top: 10, right: 20, bottom: 30, left: 40 });
+     * ```
+     */
+    setPopupPadding(padding?: PaddingOptions): this {
```
Comment (HarelM) [inline src/ui/popup.ts]: Same here - setPadding

File: src/ui/popup.test.ts [hunk 4]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+        });
+
+        test('setPopupPadding updates options and triggers update', () => {
+            const map = createMap();
+            const popup = new Popup()
+                .setText('Test')
+                .setLngLat([0, 0])
+                .addTo(map);
+
+            const updateSpy = vi.spyOn(popup, '_update');
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Is there a more "public" API to check this?

File: src/ui/popup.test.ts [hunk 5]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+                .setLngLat(nearCornerLngLat)
+                .addTo(map);
+
+            // Get the anchor class to verify positioning
+            const elementNoPadding = popupNoPadding.getElement();
+            const initialAnchor = Array.from(elementNoPadding.classList).find(cls => cls.includes('anchor'));
+
+            popupNoPadding.remove();
+
+            // Create popup with padding
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Should this test be split?

File: src/ui/popup.test.ts [hunk 6]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+            const popup = new Popup()
+                .setText('Test')
+                .setLngLat([0, 0])
+                .addTo(map);
+
+            popup.setPopupPadding({top: 5, right: 10});
+            expect(popup.options.popupPadding).toEqual({top: 5, right: 10});
+        });
+
+        test('setPopupPadding accepts null/undefined to clear padding', () => {
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Split null and undefined to 2 tests.

File: src/ui/popup.test.ts [hunk 7]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+            popupNoPadding.remove();
+            popupWithPadding.remove();
+        });
+
+        test('offset and padding interaction preserves offset behavior', () => {
+            const map = createMap();
+
+            const offset = {top: [0, -20], bottom: [0, 20], left: [20, 0], right: [-20, 0]} as any;
+
+            // Test that offset is still applied correctly when padding is present
```
Comment (HarelM) [inline src/ui/popup.test.ts]: I think this comment is redundant given the good test name you wrote.

File: src/ui/popup.test.ts [hunk 8]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+            const popup = new Popup({offset, popupPadding: {top: 10, right: 10, bottom: 10, left: 10}})
+                .setText('Test with offset and padding')
+                .setLngLat([0, 0])
+                .addTo(map);
+
+            expect(popup.getElement()).toBeDefined();
+            // The fact that it renders without error confirms offset handling is intact
+        });
+
+        test('edge cases - negative, zero, extreme values', () => {
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Split into different tests.

File: src/ui/popup.test.ts [hunk 9]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+
+            const edgeCases = [
+                {top: 0, right: 0, bottom: 0, left: 0},           // zero
+                {top: -10, right: -10, bottom: -10, left: -10},   // negative
+                {top: 1000000, right: 1000000, bottom: 1000000, left: 1000000}, // extremely large
+                {top: -50, right: 0, bottom: 100, left: 50},      // mixed values
+            ];
+
+            edgeCases.forEach((padding, index) => {
+                try {
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Use expect(...).not.toThrow or some similar.

File: src/ui/popup.test.ts [hunk 10]
```diff
@@ -887,4 +887,175 @@ describe('popup', () => {
...
+
+        test('trackPointer compatibility', () => {
+            const map = createMap();
+
+            const popup = new Popup({popupPadding: {top: 20, right: 20, bottom: 20, left: 20}})
+                .setText('Track pointer test')
+                .trackPointer()
+                .addTo(map);
+
+            // Should not crash and should have track-pointer class
```
Comment (HarelM) [inline src/ui/popup.test.ts]: Better to move this comment to the test name, it's more informative.

**Comment by HarelM** [comment]
Thanks for pushing this through! 
I've added some comments, but looks good otherwise.

File: src/ui/popup.ts (hunk 1 above)
Comment (lucaswoj) [reply src/ui/popup.ts]: Done in e0694f80d

File: src/ui/popup.ts (hunk 2 above)
Comment (lucaswoj) [reply src/ui/popup.ts]: Fixed in bd7cedd21

File: src/ui/popup.ts (hunk 3 above)
Comment (lucaswoj) [reply src/ui/popup.ts]: Done in 1b22232d4

File: src/ui/popup.test.ts (hunk 4 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: I'm not aware of a better way to write this test. It was copied over verbatim from #6052. We can remove if you prefer.

File: src/ui/popup.test.ts (hunk 5 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in 187ee5883

File: src/ui/popup.test.ts (hunk 6 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in 85a28f8e497220bdd9e4873e5b3c4aae1eab1346

File: src/ui/popup.test.ts (hunk 7 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in d41d3a470

File: src/ui/popup.test.ts (hunk 8 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in cbd433931

File: src/ui/popup.test.ts (hunk 9 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Unnecessary as of cbd4339316539424ac572c12e11ba6ab11aaeb6b. Thrown errors cause tests to fail.

File: src/ui/popup.test.ts (hunk 10 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in 946cab610

**Comment by lucaswoj** [comment]
This isn't working for me in manual testing -- hold off on merging for a moment.

**Comment by lucaswoj** [comment]
Dec-18-2025 15-21-46

False alarm, this is working great, I was just testing against the wrong build.

File: src/ui/popup.test.ts (hunk 4 above)
Comment (HarelM) [reply src/ui/popup.test.ts]: Lets remove it then.

**Comment by HarelM** [comment]
Thanks for all the work on this! One final comment before merging. 
As a side note, do you think getPadding is needed?

File: src/ui/popup.test.ts (hunk 4 above)
Comment (lucaswoj) [reply src/ui/popup.test.ts]: Done in 4c097e753

**Comment by lucaswoj** [comment]
> Thanks for all the work on this! One final comment before merging. As a side note, do you think getPadding is needed?

No. The precedent appears to be: anything that can be read from `Popup#options` doesn't have a dedicated getter. For example, there's no `getOffset` method because you can just read `popup.options.offset`. If you'd like to revisit the design of the Popup API more fundamentally, that belongs in a separate PR.

**Review by HarelM — APPROVED** [review]
